### PR TITLE
feat: add config command

### PR DESCRIPTION
**Pull request by LexLuthr**
Fixes https://github.com/filecoin-project/boost/issues/1488

```
# boostd config
NAME:
   boostd config - Display Boost config

USAGE:
   boostd config command [command options] [arguments...]

COMMANDS:
   default  Print default node config
   updated  Print updated node config
   help, h  Shows a list of commands or help for one command

OPTIONS:
   --help, -h  show help
root@3d7cba20624d:/app# boostd config default
# The version of the config file (used for migrations)
#
# type: int
# env var: BOOST__CONFIGVERSION
#ConfigVersion = 0
```

### CODE REVIEW

**Comment by LexLuthr** [comment]
boostd config updated --no-comment
```
ConfigVersion = 4
SealerApiInfo = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.0LyVxqOde8UjLTcHPEo3VEBILtPQqCDNEHcoCbTRQ_Y:/ip4/127.0.0.1/tcp/2345/http"
SectorIndexApiInfo = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.0LyVxqOde8UjLTcHPEo3VEBILtPQqCDNEHcoCbTRQ_Y:/ip4/127.0.0.1/tcp/2345/http"

[API]
  ListenAddress = "/ip4/127.0.0.1/tcp/1288/http"
  RemoteListenAddress = ""
  Timeout = "30s"

[Backup]
  DisableMetadataLog = false

[Libp2p]
  ListenAddresses = ["/ip4/0.0.0.0/tcp/50000", "/ip6/::/tcp/0"]
  AnnounceAddresses = []
  NoAnnounceAddresses = []
  DisableNatPortMap = false
  ConnMgrLow = 150
  ConnMgrHigh = 180
  ConnMgrGrace = "20s"

[Pubsub]
  Bootstrapper = false
  RemoteTracer = ""
  JsonTracer = ""
  ElasticSearchTracer = ""
  ElasticSearchIndex = ""
  TracerSourceAuth = ""

[Storage]
  ParallelFetchLimit = 10
  StorageListRefreshDuration = "1h0m0s"
  RedeclareOnStorageListRefresh = true

[Dealmaking]
  ConsiderOnlineStorageDeals = true
  ConsiderOfflineStorageDeals = true
  ConsiderOnlineRetrievalDeals = true
  ConsiderOfflineRetrievalDeals = true
  ConsiderVerifiedStorageDeals = true
  ConsiderUnverifiedStorageDeals = true
  PieceCidBlocklist = []
  ExpectedSealDuration = "24h0m0s"
  MaxDealStartDelay = "336h0m0s"
  MaxProviderCollateralMultiplier = 2
  MaxStagingDealsBytes = 2000000000
  MaxStagingDealsPercentPerHost = 0
  StartEpochSealingBuffer = 480
  DealProposalLogDuration = "24h0m0s"
  RetrievalLogDuration = "24h0m0s"
  StalledRetrievalTimeout = "30s"
  Filter = ""
  RetrievalFilter = ""
  BlockstoreCacheMaxShards = 20
  BlockstoreCacheExpiry = "30s"
  IsUnsealedCacheExpiry = "5m0s"
  MaxTransferDuration = "24h0m0s"
  RemoteCommp = false
  MaxConcurrentLocalCommp = 1
  HTTPRetrievalMultiaddr = ""
  HttpTransferMaxConcurrentDownloads = 20
  HttpTransferStallCheckPeriod = "30s"
  HttpTransferStallTimeout = "5m0s"
  BitswapPeerID = ""
  BitswapPrivKeyFile = ""
  DealLogDurationDays = 30
  SealingPipelineCacheTimeout = "30s"
  FundsTaggingEnabled = true
  [Dealmaking.RetrievalPricing]
    Strategy = "default"
    [Dealmaking.RetrievalPricing.Default]
      VerifiedDealsFreeTransfer = true
    [Dealmaking.RetrievalPricing.External]
      Path = ""

[Wallets]
  Miner = "t01000"
  PublishStorageDeals = "t3rjh3byfhhrlmdl6ofou544rmwuazp4bbqzuleahtn66ejnh73bsaonwyg54qrokvnbv5bmg37dfd5vu2bx4q"
  DealCollateral = "t3qa5i54vprhl7gmdfgriubvjc5kdffx7cvbqvumopgvsy2osre5dcjdjhz5coruxp4co3chj3gbtxin7vtoia"
  PledgeCollateral = ""

[Graphql]
  ListenAddress = "127.0.0.1"
  Port = 8080

[Tracing]
  Enabled = false
  ServiceName = "boostd"
  Endpoint = ""

[ContractDeals]
  Enabled = false
  AllowlistContracts = []
  From = "0x0000000000000000000000000000000000000000"

[LotusDealmaking]
  ConsiderOnlineStorageDeals = true
  ConsiderOfflineStorageDeals = true
  ConsiderOnlineRetrievalDeals = true
  ConsiderOfflineRetrievalDeals = true
  ConsiderVerifiedStorageDeals = true
  ConsiderUnverifiedStorageDeals = true
  PieceCidBlocklist = []
  ExpectedSealDuration = "24h0m0s"
  MaxDealStartDelay = "336h0m0s"
  PublishMsgPeriod = "1h0m0s"
  MaxDealsPerPublishMsg = 8
  MaxProviderCollateralMultiplier = 2
  MaxStagingDealsBytes = 0
  SimultaneousTransfersForStorage = 20
  SimultaneousTransfersForStoragePerClient = 0
  SimultaneousTransfersForRetrieval = 20
  StartEpochSealingBuffer = 480
  Filter = ""
  RetrievalFilter = ""
  [LotusDealmaking.RetrievalPricing]
    Strategy = "default"
    [LotusDealmaking.RetrievalPricing.Default]
      VerifiedDealsFreeTransfer = true
    [LotusDealmaking.RetrievalPricing.External]
      Path = ""

[LotusFees]
  MaxPublishDealsFee = "0.05 FIL"
  MaxMarketBalanceAddFee = "0.007 FIL"

[DAGStore]
  RootDir = ""
  MaxConcurrentIndex = 5
  MaxConcurrentReadyFetches = 0
  MaxConcurrentUnseals = 0
  MaxConcurrencyStorageCalls = 100
  GCInterval = "1m0s"

[IndexProvider]
  Enable = true
  EntriesCacheCapacity = 1024
  EntriesChunkSize = 16384
  TopicName = ""
  PurgeCacheOnStart = false
  [IndexProvider.Announce]
    AnnounceOverHttp = false
    DirectAnnounceURLs = ["https://cid.contact/ingest/announce"]
  [IndexProvider.HttpPublisher]
    Enabled = false
    PublicHostname = ""
    Port = 3104
```



lexluthr@Lexs-Mac boost % ./boostd config default --no-comment
```
ConfigVersion = 0
SealerApiInfo = ""
SectorIndexApiInfo = ""

[API]
  ListenAddress = "/ip4/127.0.0.1/tcp/1288/http"
  RemoteListenAddress = ""
  Timeout = "30s"

[Backup]
  DisableMetadataLog = false

[Libp2p]
  ListenAddresses = ["/ip4/0.0.0.0/tcp/0", "/ip6/::/tcp/0"]
  AnnounceAddresses = []
  NoAnnounceAddresses = []
  DisableNatPortMap = false
  ConnMgrLow = 150
  ConnMgrHigh = 180
  ConnMgrGrace = "20s"

[Pubsub]
  Bootstrapper = false
  RemoteTracer = ""
  JsonTracer = ""
  ElasticSearchTracer = ""
  ElasticSearchIndex = ""
  TracerSourceAuth = ""

[Storage]
  ParallelFetchLimit = 10
  StorageListRefreshDuration = "1h0m0s"
  RedeclareOnStorageListRefresh = true

[Dealmaking]
  ConsiderOnlineStorageDeals = true
  ConsiderOfflineStorageDeals = true
  ConsiderOnlineRetrievalDeals = true
  ConsiderOfflineRetrievalDeals = true
  ConsiderVerifiedStorageDeals = true
  ConsiderUnverifiedStorageDeals = true
  PieceCidBlocklist = []
  ExpectedSealDuration = "24h0m0s"
  MaxDealStartDelay = "336h0m0s"
  MaxProviderCollateralMultiplier = 2
  MaxStagingDealsBytes = 0
  MaxStagingDealsPercentPerHost = 0
  StartEpochSealingBuffer = 480
  DealProposalLogDuration = "24h0m0s"
  RetrievalLogDuration = "24h0m0s"
  StalledRetrievalTimeout = "30s"
  Filter = ""
  RetrievalFilter = ""
  BlockstoreCacheMaxShards = 20
  BlockstoreCacheExpiry = "30s"
  IsUnsealedCacheExpiry = "5m0s"
  MaxTransferDuration = "24h0m0s"
  RemoteCommp = false
  MaxConcurrentLocalCommp = 1
  HTTPRetrievalMultiaddr = ""
  HttpTransferMaxConcurrentDownloads = 20
  HttpTransferStallCheckPeriod = "30s"
  HttpTransferStallTimeout = "5m0s"
  BitswapPeerID = ""
  BitswapPrivKeyFile = ""
  DealLogDurationDays = 30
  SealingPipelineCacheTimeout = "30s"
  FundsTaggingEnabled = true
  [Dealmaking.RetrievalPricing]
    Strategy = "default"
    [Dealmaking.RetrievalPricing.Default]
      VerifiedDealsFreeTransfer = true
    [Dealmaking.RetrievalPricing.External]
      Path = ""

[Wallets]
  Miner = ""
  PublishStorageDeals = ""
  DealCollateral = ""
  PledgeCollateral = ""

[Graphql]
  ListenAddress = "127.0.0.1"
  Port = 8080

[Tracing]
  Enabled = false
  ServiceName = "boostd"
  Endpoint = ""

[ContractDeals]
  Enabled = false
  AllowlistContracts = []
  From = "0x0000000000000000000000000000000000000000"

[LotusDealmaking]
  ConsiderOnlineStorageDeals = true
  ConsiderOfflineStorageDeals = true
  ConsiderOnlineRetrievalDeals = true
  ConsiderOfflineRetrievalDeals = true
  ConsiderVerifiedStorageDeals = true
  ConsiderUnverifiedStorageDeals = true
  PieceCidBlocklist = []
  ExpectedSealDuration = "24h0m0s"
  MaxDealStartDelay = "336h0m0s"
  PublishMsgPeriod = "1h0m0s"
  MaxDealsPerPublishMsg = 8
  MaxProviderCollateralMultiplier = 2
  MaxStagingDealsBytes = 0
  SimultaneousTransfersForStorage = 20
  SimultaneousTransfersForStoragePerClient = 0
  SimultaneousTransfersForRetrieval = 20
  StartEpochSealingBuffer = 480
  Filter = ""
  RetrievalFilter = ""
  [LotusDealmaking.RetrievalPricing]
    Strategy = "default"
    [LotusDealmaking.RetrievalPricing.Default]
      VerifiedDealsFreeTransfer = true
    [LotusDealmaking.RetrievalPricing.External]
      Path = ""

[LotusFees]
  MaxPublishDealsFee = "0.05 FIL"
  MaxMarketBalanceAddFee = "0.007 FIL"

[DAGStore]
  RootDir = ""
  MaxConcurrentIndex = 5
  MaxConcurrentReadyFetches = 0
  MaxConcurrentUnseals = 0
  MaxConcurrencyStorageCalls = 100
  GCInterval = "1m0s"

[IndexProvider]
  Enable = true
  EntriesCacheCapacity = 1024
  EntriesChunkSize = 16384
  TopicName = ""
  PurgeCacheOnStart = false
  [IndexProvider.Announce]
    AnnounceOverHttp = false
    DirectAnnounceURLs = ["https://cid.contact/ingest/announce"]
  [IndexProvider.HttpPublisher]
    Enabled = false
    PublicHostname = ""
    Port = 3104
```

**Comment by dirkmc** [comment]
The config diff option is a nice idea 👍 

Is it possible to remove the empty sections? eg these parts:
```
[Graphql]
[Tracing]
[ContractDeals]
[LotusDealmaking]
  [LotusDealmaking.RetrievalPricing]
    [LotusDealmaking.RetrievalPricing.Default]
    [LotusDealmaking.RetrievalPricing.External]
```

**Comment by dirkmc** [comment]
I'm not sure if this is going to work if the command is run from a machine that is different from the machine where the config file is.. I think probably we need the command to invoke the RPC API

File: node/config/update.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/BurntSushi/toml"
+)
+
+// ConfigUpdate takes in a config and a default config and optionally comments out default values
```
Comment (dirkmc) [inline node/config/update.go:14]: It looks like there is a lot of shared code between ConfigUpdate and ConfigDiff.
Can we refactor out the common code into functions called from ConfigUpdate and ConfigDiff?

**Comment by LexLuthr** [comment]
Latest output:

./boostd config updated --diff --no-comment        
```                  
ConfigVersion = 4
SealerApiInfo = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.0LyVxqOde8UjLTcHPEo3VEBILtPQqCDNEHcoCbTRQ_Y:/ip4/127.0.0.1/tcp/2345/http"
SectorIndexApiInfo = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.0LyVxqOde8UjLTcHPEo3VEBILtPQqCDNEHcoCbTRQ_Y:/ip4/127.0.0.1/tcp/2345/http"
[Libp2p]
  ListenAddresses = ["/ip4/0.0.0.0/tcp/50000", "/ip6/::/tcp/0"]

[Dealmaking]
  MaxStagingDealsBytes = 2000000000

[Wallets]
  Miner = "t01000"

  PublishStorageDeals = "t3rjh3byfhhrlmdl6ofou544rmwuazp4bbqzuleahtn66ejnh73bsaonwyg54qrokvnbv5bmg37dfd5vu2bx4q"

  DealCollateral = "t3qa5i54vprhl7gmdfgriubvjc5kdffx7cvbqvumopgvsy2osre5dcjdjhz5coruxp4co3chj3gbtxin7vtoia"
```

./boostd config updated --diff             
```
# The version of the config file (used for migrations)
#
# type: int
ConfigVersion = 4
# The connect string for the sealing RPC API (lotus miner)
#
# type: string
SealerApiInfo = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.0LyVxqOde8UjLTcHPEo3VEBILtPQqCDNEHcoCbTRQ_Y:/ip4/127.0.0.1/tcp/2345/http"
# The connect string for the sector index RPC API (lotus miner)
#
# type: string
SectorIndexApiInfo = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.0LyVxqOde8UjLTcHPEo3VEBILtPQqCDNEHcoCbTRQ_Y:/ip4/127.0.0.1/tcp/2345/http"
[Libp2p]
  # Binding address for the libp2p host - 0 means random port.
  # Format: multiaddress; see https://multiformats.io/multiaddr/
  #
  # type: []string
  ListenAddresses = ["/ip4/0.0.0.0/tcp/50000", "/ip6/::/tcp/0"]

[Dealmaking]
  # The maximum allowed disk usage size in bytes of downloaded deal data
  # that has not yet been passed to the sealing node by boost.
  # When the client makes a new deal proposal to download data from a host,
  # boost checks this config value against the sum of:
  # - the amount of data downloaded in the staging area
  # - the amount of data that is queued for download
  # - the amount of data in the proposed deal
  # If the total amount would exceed the limit, boost rejects the deal.
  # Set this value to 0 to indicate there is no limit.
  #
  # type: int64
  MaxStagingDealsBytes = 2000000000

[Wallets]
  # The "owner" address of the miner
  #
  # type: string
  Miner = "t01000"

  # The wallet used to send PublishStorageDeals messages.
  # Must be a control or worker address of the miner.
  #
  # type: string
  PublishStorageDeals = "t3rjh3byfhhrlmdl6ofou544rmwuazp4bbqzuleahtn66ejnh73bsaonwyg54qrokvnbv5bmg37dfd5vu2bx4q"

  # The wallet used as the source for storage deal collateral
  #
  # type: string
  DealCollateral = "t3qa5i54vprhl7gmdfgriubvjc5kdffx7cvbqvumopgvsy2osre5dcjdjhz5coruxp4co3chj3gbtxin7vtoia"
```

**Comment by dirkmc** [comment]
Looks great 👍 
My only suggestion is to make an RPC call (like the lotus version of this command) rather than operating directly on the config file locally.

**Comment by dirkmc** [comment]
Nevermind I see that the lotus version also operates on the local file, so let's keep that behaviour.

**Review by dirkmc — APPROVED** [review]
Nice work 🙌

**Pull request merged**